opm/upscaling/RelPermUtils.hpp

```diff
@@ -0,0 +1,161 @@
+
+      //! \brief Collect results from all MPI nodes.
+      void collectResults();
+
+      //! \brief Calculate relperm values from phase permeabilities.
+      //! \param[in] phase The phase to calculate values for (0-indexed).
+      //! \return The phase permeability values.
+      std::vector<std::vector<double>> getRelPerm(int phase) const;
+
+      //! \brief Upscale single phase permeability
+      void upscaleSinglePhasePermeability();
+
+      //! \brief Do sanity checks for input file.
+      //! \param[in] deck The deck to sanity check.
+      //! \param[in] minPerm Minimum permeability.
+      //! \param[in] maxPerm Maximum permeability.
+      //! \param[in] minPoro Minimum porosity.
+      //! \details Throws error string.
+      void sanityCheckInput(Opm::DeckConstPtr deck,
+                            double minPerm, double maxPerm, double minPoro);
+
+      //! \brief Check that input relperm curevs specify critical saturations.
+      void checkCriticalSaturations();
+
+      //! \brief Setup requested boundary conditions.
+      //! \details Uses the following options: bc
+      void setupBoundaryConditions();
+
+      //! \brief Tesselate grid
+      //! \param[in] deck The grid to tesselate.
+      //! \param[in] options Option structure.
+      //! \details Uses the following options: linsolver_tolerance,
+      //!          linsolver_verbosity, linsolver_type, linsolver_max_iterations,
+      //!          linsolver_smooth_steps, linsolver_prolongate_factor, minPerm
+      //! \return Time used for tesselation.
+      double tesselateGrid(Opm::DeckConstPtr deck);
+
+      //! \brief Find cell center pressure gradient for every cell.
+      //! \param[in] res The number of cells in each direction.
+      //! \details Uses the following options: gravity, waterDensity, oilDensity
+      void calculateCellPressureGradients(const std::array<int,3>& res);
+
+      //! \brief Calculate minimum and maximum capillary pressures.
+      //! \details Uses the following options: maxPermContrast, minPerm,
+      //!                                      gravity, linsolver_tolerance
+      void calculateMinMaxCapillaryPressure();
+
+      //! \brief Upscale capillary pressure.
+      //! \details Uses the following options: saturationThreshold
+      void upscaleCapillaryPressure();
+
+      //! \brief Upscale permeabilities.
+      //! \param[in] mpi_rank MPI rank of this process.
+      //! \details Uses the following options:  minPerm, maxPermContrast
+      //! \return Tuple with (total time, time per point).
+      std::tuple<double,double> upscalePermeability(int mpi_rank);
+    private:
+      //! \brief Perform critical saturation check for a single curve.
+      //! \param[in,out] func Function to check for.
+      //! \details Checks that minimum relperm is larger than the threshold.
+      //! \return True if test passed, false otherwise.
+      bool checkCurve(MonotCubicInterpolator& func);
```

Comment:
no, if the minimum value is below the critical, the function is updated 
with a constant value of 0.